cmd/virt-launcher/virt-launcher.go

```diff
@@ -510,55 +510,37 @@ func ForkAndMonitor(containerDiskDir string) (int, error) {
 		return 1, err
 	}
 
-	exitStatus := make(chan syscall.WaitStatus, 10)
-	sigs := make(chan os.Signal, 10)
-	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD)
-	go func() {
-		for sig := range sigs {
-			switch sig {
-			case syscall.SIGCHLD:
-				var wstatus syscall.WaitStatus
-				wpid, err := syscall.Wait4(-1, &wstatus, syscall.WNOHANG, nil)
-				if err != nil {
-					log.Log.Reason(err).Errorf("Failed to reap process %d", wpid)
-				}
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGCHLD)
 
-				// there's a race between cmd.Wait() and syscall.Wait4 when
-				// cleaning up the cmd's pid after it exits. This allows us
-				// to detect the correct exit code regardless of which wait
-				// wins the race.
-				if wpid == cmd.Process.Pid {
-					exitStatus <- wstatus
-				}
+	// wait for virt-launcher and collect the exit code
+	exitCode := 0
+	virtLauncherDone := false
+	for !virtLauncherDone {
+		switch sig := <-sigChan; sig {
+		case syscall.SIGCHLD:
+			var wstatus syscall.WaitStatus
+			wpid, err := syscall.Wait4(-1, &wstatus, syscall.WNOHANG, nil)
+			if err != nil {
+				log.Log.Reason(err).Errorf("Failed to reap process %d", wpid)
+			}
 
-			default:
-				log.Log.V(3).Log("signalling virt-launcher to shut down")
-				err := cmd.Process.Signal(syscall.SIGTERM)
-				sig.Signal()
-				if err != nil {
-					log.Log.Reason(err).Errorf("received signal %s but can't signal virt-launcher to shut down", sig.String())
-				}
+			if wpid == cmd.Process.Pid {
+				exitCode = wstatus.ExitStatus()
+				virtLauncherDone = true
 			}
-		}
-	}()
 
-	// wait for virt-launcher and collect the exit code
-	exitCode := 0
-	if err := cmd.Wait(); err != nil {
-		select {
-		case status := <-exitStatus:
-			exitCode = int(status)
 		default:
-			exitCode = 1
-			if exiterr, ok := err.(*exec.ExitError); ok {
-				if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
-					exitCode = status.ExitStatus()
-				}
+			log.Log.V(3).Log("signalling virt-launcher to shut down")
+			err := cmd.Process.Signal(syscall.SIGTERM)
+			sig.Signal()
```

Comment:
Is this line necessary? I see in the documentation that:
```
type Signal interface {
	String() string
	Signal() // to distinguish from other Stringers
}
```

And going into the source code it seems to be a no-op (in `syscall_unix.go`):
```
func (s Signal) Signal() {}
```

Reply:
I guess you're right, it seems indeed like a no-op. @rmohr I see git blames you :) any reason to keep this line?

Reply:
/unhold

I'm removing the hold label, we can delete this line in a follow up PR.